runtime/tests/checker/type_inference_test.go

```diff
@@ -0,0 +1,293 @@
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: [[[Int8]]] = [[[1, 2, 3], [4]], []]
+			}
+		`)
+
+		require.NoError(t, err)
+	})
+}
+
+func TestDictionaryTypeInference(t *testing.T) {
+
+	t.Parallel()
+
+	t.Run("numerics", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: {Int8:Int64} = {0: 6, 1: 5}
+				var y: {Int8:Int64?} = {0: 6, 1: 5}
+			}
+		`)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("heterogeneous", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: {Int:AnyStruct} = {0: 6, 1: "hello", 2: nil}
+			}
+		`)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("nested", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: {Int:{Int:{Int:AnyStruct}}} = {0: {0: {0: 6}, 1: {0: 7}}}
+			}
+		`)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("invalid", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: {Int8:Int64} = {23423:6, 1:5}
+			}
+		`)
+
+		require.Error(t, err)
+		assert.IsType(t, &sema.CheckerError{}, err)
+	})
+
+	t.Run("nested invalid", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: {Int:{Int:{Int:Int8}}} = {0: {0: {0: "hello"}, 1: {0: 7}}}
+			}
+		`)
+
+		require.Error(t, err)
+		assert.IsType(t, &sema.CheckerError{}, err)
+	})
+}
+
+func TestReturnTypeInference(t *testing.T) {
+
+	t.Parallel()
+
+	_, err := ParseAndCheck(t, `
+		fun test(): [Int8] {
+			return [1, 2, 3]
+		}
+	`)
+
+	require.NoError(t, err)
+}
+
+func TestFunctionArgumentTypeInference(t *testing.T) {
+
+	t.Parallel()
+
+	_, err := ParseAndCheck(t, `
+		fun test() {
+			foo(a: [1, 2, 3])
+		}
+
+		fun foo(a: [Int8]) {
+		}
+	`)
+
+	// Type inferring for function arguments is not supported yet.
+	require.Error(t, err)
+	assert.IsType(t, &sema.CheckerError{}, err)
+}
+
+func TestBinaryExpressionTypeInference(t *testing.T) {
+
+	t.Parallel()
+
+	t.Run("integer add", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: Int8 = 5 + 6
+			}
+		`)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("fixed point add", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: Fix64 = 5.0 + 6.0
+			}
+		`)
+
+		require.NoError(t, err)
+	})
+}
+
+func TestUnaryExpressionTypeInference(t *testing.T) {
+
+	t.Parallel()
+
+	t.Run("invalid negate", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var b : Bool =  !"string"
+			}
+		`)
+
+		require.Error(t, err)
+		require.IsType(t, &sema.CheckerError{}, err)
+
+		checkerErr := err.(*sema.CheckerError)
+
+		// Should only give a single error
+		require.Len(t, checkerErr.Errors, 1)
+
+		assert.IsType(t, &sema.InvalidUnaryOperandError{}, checkerErr.Errors[0])
+	})
+}
+
+func TestForceExpressionTypeInference(t *testing.T) {
+
+	t.Parallel()
+
+	t.Run("array forced", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: [Int8] = [5, 7, 2]!
```

Comment:
Nice, this answers the question I had above regarding force-unwrapping non-optionals 👍